### ISSUE TITLE
negative loss in tf version

### ISSUE THREAD

**Issue by imodile** [HawkAaron/warp-transducer#21]
hi,
I have tried binding rnnt loss to tensorflow 1.4 and by testing the loss function I got negative results. The whole code are as follows. Why did it happen?

import numpy as np
import tensorflow as tf
from warprnnt_tensorflow import rnnt_loss

B = 1; T = 3; U = 2; V = 5; blank = V

logits = tf.placeholder(tf.float32, [None, T, U, V])
labels = tf.placeholder(tf.int32, [None, U])
input_length = tf.placeholder(tf.int32, [None])
label_length = tf.placeholder(tf.int32, [None])

logits = tf.nn.log_softmax(logits)
costs = rnnt_loss(logits, labels, input_length, label_length, blank)
grad = tf.gradients(costs, [logits])

a = np.array(
                           [[
                            [[0.1, 0.6, 0.1, 0.1, 0.1],
                             [0.1, 0.1, 0.6, 0.1, 0.1]],
                               
                            [[0.1, 0.6, 0.1, 0.1, 0.1],
                             [0.1, 0.1, 0.6, 0.1, 0.1]],
                            
                            [[0.1, 0.6, 0.1, 0.1, 0.1],
                             [0.1, 0.3, 0.2, 0.3, 0.1]],
                           
                           ]]
    , dtype=np.float32)

c = np.array([3], dtype=np.int32)
print(a.shape)
print(c)

b = np.array([[0,5]], dtype=np.int32)
d = np.array([1], dtype=np.int32)
print(b.shape)
print(d)

feed = {logits: a, labels: b, input_length: c, label_length: d}
with tf.Session() as sess:
    logit, cost, grads = sess.run([logits, costs, grad], feed_dict=feed)
    print(cost)

And does rnnt_loss computing process include softmax function? So when I get logits after rnn or fully connected layer, do I need to softmax the logits?

Thanks a lot!

**Comment by HawkAaron** [comment]
@imodile For the cpu op, you need to calculate log_softmax manually; the gpu op already done log_softmax inside.

**Comment by imodile** [comment]
hi, thanks for your reply!

Another question is about the negative loss from the above example, why does that happen? 
I am really confused and will really appreciate your explanation!

**Comment by HawkAaron** [comment]
Everything works well in my environment, please rebuild this package and run all the test.

**Comment by HawkAaron** [comment]
I've updated the code, maybe you could try it again.
